src/libsystemd/sd-json/json-util.c

```diff
@@ -144,3 +146,138 @@ int json_dispatch_path(const char *name, sd_json_variant *variant, sd_json_dispa
+
+        if (sd_json_variant_is_null(variant)) {
+                pidref_done(p);
+                return 0;
+        }
+
+        struct {
+                uint64_t pid, fd_id;
+                sd_id128_t boot_id;
+        } data = {};
+
+        if (sd_json_variant_is_integer(variant))
+                /* Permit a simple classic integer based format */
+                data.pid = sd_json_variant_integer(variant);
+        else if (sd_json_variant_is_string(variant)) {
+                /* As usual, allow integers be encoded as strings too */
+                r = safe_atou64(sd_json_variant_string(variant), &data.pid);
+                if (r < 0)
+                        return json_log(variant, flags, r, "JSON field '%s' is not a numeric PID.", strna(name));
+        } else if (sd_json_variant_is_object(variant)) {
+
+                static const sd_json_dispatch_field dispatch_table[] = {
+                        { "pid",     _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint64, voffsetof(data, pid),     SD_JSON_MANDATORY },
+                        { "pidfdId", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint64, voffsetof(data, fd_id),   0                 },
+                        { "bootId",  SD_JSON_VARIANT_STRING,        sd_json_dispatch_id128,  voffsetof(data, boot_id), 0                 },
+                        {}
+                };
+
+                r = sd_json_dispatch(variant, dispatch_table, flags, &data);
+                if (r < 0)
+                        return r;
+        } else
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is neither a numeric PID or a PID object.", strna(name));
+
+        /* Before casting the 64bit data.pid field to pid_t, let's ensure it fits the pid_t range. */
+        if (data.pid > PID_T_MAX || !pid_is_valid(data.pid))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is does not contain a valid PID.", strna(name));
+
+        int local_boot_id = -1; /* tristate */
+        if (!sd_id128_is_null(data.boot_id)) {
+                sd_id128_t my_boot_id;
+
+                r = sd_id128_get_boot(&my_boot_id);
+                if (r < 0) {
+                        json_log(variant, flags | (FLAGS_SET(flags, SD_JSON_STRICT) ? 0 : SD_JSON_DEBUG), r, "Unable to get local boot ID to validate JSON field '%s': %m", strna(name));
+                        if (FLAGS_SET(flags, SD_JSON_STRICT))
+                                return r;
+                } else {
+                        local_boot_id = sd_id128_equal(data.boot_id, my_boot_id);
+                        if (!local_boot_id) {
+                                json_log(variant, flags | (FLAGS_SET(flags, SD_JSON_STRICT) ? 0 : SD_JSON_DEBUG), 0, "JSON field '%s' refers to non-local PID.", strna(name));
+                                if (FLAGS_SET(flags, SD_JSON_STRICT))
+                                        return -ESRCH;
```

Comment:
Hmm, it's not clear to me that we should ever store non-local pids in a PidRef. So far it's designed to reference something we track locally, plus without boot id the resulting PidRef cannot be used to refer to anything sanely again.

Reply:
Well, I am thinking of a client here that talks to a remote service that returns the triplet here. I think if SD_JSON_STRICT is off we should return *something*, and not fail deserializion. Hence store the pid and the pidid in the PidRef, but do not acquire a pidfd in that case, because that makes no sense.

Of course, let's hope the client doesn't then make use of the pid in any other way than for display purposes, but i think that shouldn't stop us from deserializing this.

I was tempted to maybe extend PidRef with a boot id, so that we can pass this stuff around losslessly, but it sounded like overkill to me to waste another 16byte on this to me... dunno. certainly a discussion to be had, but probably separately from this here.

Reply:
what about this, to address the issue: #34719 what do you think?